engine/src/rocksdb_engine.rs

```diff
@@ -423,52 +422,54 @@ impl RocksSnapshot {
     }
 
     /// Write snapshot data
-    #[allow(clippy::indexing_slicing)] // safe indexing
-    async fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        if buf.is_empty() {
-            return Ok(0);
-        }
-
-        let (mut next_buf, mut written_bytes) = (buf, 0);
+    async fn write(&mut self, buf: &mut Bytes) -> io::Result<()> {
         if self.meta.is_current {
-            let meta_len_bytes: [u8; 8] = next_buf[0..8]
-                .try_into()
-                .map_err(|e| io::Error::new(ErrorKind::Other, e))?;
-            let meta_len = u64::from_le_bytes(meta_len_bytes);
+            if buf.len() < 8 {
+                return Err(io::Error::new(
+                    ErrorKind::UnexpectedEof,
+                    "cannot read meta length from buffer",
+                ));
+            }
+            let meta_len = buf.get_u64_le();
 
-            let meta_bytes: Vec<u8> = next_buf[8..meta_len.overflow_add(8).numeric_cast()]
-                .try_into()
-                .unwrap_or_else(|_e| unreachable!("infallible"));
+            if buf.len() < meta_len.numeric_cast() {
+                return Err(io::Error::new(
+                    ErrorKind::UnexpectedEof,
+                    "cannot read meta from buffer",
+                ));
+            };
+            let meta_bytes = buf.split_to(meta_len.numeric_cast());
             let meta = bincode::deserialize(&meta_bytes)
                 .map_err(|e| io::Error::new(ErrorKind::Other, e))?;
 
             self.apply_snap_meta(meta);
             let mut data = Vec::new();
-            data.extend(meta_len_bytes);
+            data.extend(meta_len.to_le_bytes());
```

Comment:
Shall we change the `self.meta.data` from `Cursor<Vec<u8>>` to `Cursor<Bytes>`? So we can use `chain` to link `meta_len` and `meta_bytes` as a new `Bytes` without copying meta_bytes.

Reply:
fix this in https://github.com/datenlord/Xline/pull/264/commits/8626b05398cc93a43a67978e515212624ae93f71